tests/git/utils.ts

```diff
@@ -1,72 +1,270 @@
-      write: true,
+  for (const { message, files } of commits) {
+    await Promise.all(
+      files.map(({ name, contents }) =>
+        fs.promises.writeFile(path.join(gitDirs.dir, name), contents),
+      ),
+    );
+    await git.add({
+      ...gitDirs,
+      filepath: files.map(({ name }) => name),
     });
-    if (indexFile) {
-      await git.indexPack({
-        fs: efs,
-        dir: '.',
-        filepath: path.join('.git', 'objects', 'pack', pack.filename),
-      });
+    await git.commit({
+      ...gitDirs,
+      ...authorDetails,
+      message,
+    });
+  }
+}
+
+const objectsDirName = 'objects';
+const excludedDirs = ['pack', 'info'];
+
+/**
+ * Walks the filesystem to list out all git objects in the objects directory
+ * @param fs
+ * @param gitDir
+ */
+async function listGitObjects({
+  fs,
+  gitDir,
+}: {
+  fs: FileSystem;
+  gitDir: string;
+}) {
+  const objectsDirPath = path.join(gitDir, objectsDirName);
+  const objectSet: Set<string> = new Set();
+  const objectDirs = await fs.promises.readdir(objectsDirPath);
+  for (const objectDir of objectDirs) {
+    if (excludedDirs.includes(objectDir)) continue;
+    const objectIds = await fs.promises.readdir(
+      path.join(objectsDirPath, objectDir),
+    );
+    for (const objectId of objectIds) {
+      objectSet.add(objectDir + objectId);
     }
   }
-  return log;
+  return [...objectSet];
 }
 
-async function getPackID(efs: EncryptedFS): Promise<string> {
-  const pack = (
-    await efs.promises.readdir(path.join('.git', 'objects', 'pack'))
-  )[0];
-  return (pack as string).substring(5, 45);
+type NegotiationTestData =
+  | {
+      type: 'want';
+      objectId: string;
+      capabilityList: CapabilityList;
+    }
+  | {
+      type: 'have';
+      objectId: string;
+    }
+  | {
+      type: 'SEPARATOR' | 'done' | 'none';
+    };
+
+function generateTestNegotiationLine(data: NegotiationTestData, rest: Buffer) {
+  switch (data.type) {
+    case 'want': {
+      const line = Buffer.concat([
+        Buffer.from(data.type),
+        gitUtils.SPACE_BUFFER,
+        Buffer.from(data.objectId),
+        gitUtils.SPACE_BUFFER,
+        Buffer.from(data.capabilityList.join(gitUtils.SPACE_STRING)),
+        gitUtils.LINE_FEED_BUFFER,
+      ]);
+      return Buffer.concat([gitHttp.packetLineBuffer(line), rest]);
+    }
+    case 'have': {
+      const line = Buffer.concat([
+        Buffer.from(data.type),
+        gitUtils.SPACE_BUFFER,
+        Buffer.from(data.objectId),
+        gitUtils.LINE_FEED_BUFFER,
+      ]);
+      return Buffer.concat([gitHttp.packetLineBuffer(line), rest]);
+    }
+    case 'SEPARATOR':
+      return Buffer.concat([Buffer.from('0000'), rest]);
+    case 'done':
+      return Buffer.concat([Buffer.from('0009done\n'), rest]);
+    case 'none':
+      return Buffer.alloc(0);
+    default:
+      utils.never();
+  }
+}
+
+// Used to print out the contents of an `Buffer` iterable for testing
+async function* tapGen(
```

Comment:
Docblock format please.

Reply:
This function's name is a bit weird? Should it be expanded.

Reply:
`tapGen` has been removed.